### PR TITLE
feat: Relay experimental sweeping upgrades

### PR DESCRIPTION
**Pull request by maraisr**
This syncs up the react-relay types with that of current master@experimental.

This DOES NOT sync the relay-runtime _yet_. That will come after.

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react-relay ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-relay@latest/))
was missing the following properties:
1. ConnectionHandler
1. MutationTypes
1. RangeOperations
1. ConnectionHandler
1. MutationTypes

<details>
<summary>as well as these 1 other properties...</summary>
<p>RangeOperations</p>
</details>

## relay-runtime ([<kbd>unpkg</kbd>](https://unpkg.com/browse/relay-runtime@latest/))
was missing the following properties:
1. ReplaySubject
1. MutationTypes
1. RangeOperations
1. MutationHandlers
1. VIEWER_ID

<details>
<summary>as well as these 9 other properties...</summary>
<p>VIEWER_TYPE, PreloadableQueryRegistry, createPayloadFor3DField, RelayError, DEFAULT_HANDLE_KEY, isScalarAndEqual, recycleNodesInto, stableCopy, getFragmentIdentifier</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against fc273d986e101fed930bdc20c806150de3a12215
</p>

File: types/react-relay/lib/hooks.d.ts
Comment (edvinerikson) [inline types/react-relay/lib/hooks.d.ts:1]: It is: https://unpkg.com/browse/react-relay@0.0.0-experimental-183bdd28/lib/hooks.js

Comment (maraisr) [reply types/react-relay/lib/hooks.d.ts:1]: Sure its in the unpkg one, but that file doesnt exist in the relay source code. So not sure where that is coming from: https://github.com/facebook/relay/blob/master/packages/relay-experimental/index.js

File: types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts
Comment (edvinerikson) [inline types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts:41]: We need to tweak this type as TypeScript can't assign ConcreteRequests to it. Alternatively, we can perhaps change the consumers of this type to use a union of  `PreloadableConcreteRequest | ConcreteRequest`. What do you think?

Comment (maraisr) [reply types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts:43]: I've updated it and noted down why we need to union now

File: types/entria__relay-experimental/index.d.ts
Comment (maraisr) [inline types/entria__relay-experimental/index.d.ts:5]: @sibelius you okay with this change?

File: types/react-relay/index.d.ts
Comment (maraisr) [inline types/react-relay/index.d.ts:15]: @edvinerikson can you please confirm I've got this right, as well as just "agree" to have me add you here.

File: types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts
Comment (maraisr) [inline types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts:43]: I got rid of this  _DEPRECATED stuff. Its basically what used to be sent through `prepareEntrypoint` which isnt even exported in real relay anymore. TS got tricky introducing this union, so please agree we cool to remove it.

cc @alloy

File: types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts
Comment (maraisr) [inline types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts:219]: @edvinerikson  think we need to pass this type on as well, think this should be accessible in the `getEnvironmentProvider` cb. Except we're not passing this 👀

File: types/react-relay/lib/relay-experimental/helpers.d.ts
Comment (maraisr) [inline types/react-relay/lib/relay-experimental/helpers.d.ts:8]: I think in a future PR, we should standardize the "shape" of these files. Think relay-runtime has something similar to this. 

We can leave here for now @alloy and fix in a subsequent PR when i go through relay-runtime.

File: types/entria__relay-experimental/index.d.ts
Comment (sibelius) [reply types/entria__relay-experimental/index.d.ts:5]: I'm using ts 4

File: types/react-relay/index.d.ts
Comment (edvinerikson) [reply types/react-relay/index.d.ts:15]: Yep, this is right 👍🏼

File: types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts
Comment (edvinerikson) [reply types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts:219]: Not entirely following what you mean, can you elaborate?

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (alloy) [inline types/react-relay/test/relay-experimental-tests.tsx:1029]: Tiny OCD nit-pick 😎 

```suggestion
    const RootEntryPointComponent: EntryPointComponent<{ someQueryRef: SomeQuery }, {}, { onClick: () => void }> = ({
```

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (alloy) [inline types/react-relay/test/relay-experimental-tests.tsx:1042]:
```suggestion
    const entrypoint: EntryPoint<typeof RootEntryPointComponent, Params> = {
```

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (alloy) [inline types/react-relay/test/relay-experimental-tests.tsx:1043]:
```suggestion
        root: JSResource<typeof RootEntryPointComponent>(),
```

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (alloy) [inline types/react-relay/test/relay-experimental-tests.tsx:1086]:
```suggestion
        const entrypointA: EntryPoint<typeof RootEntryPointComponent, Params> = {
```

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (alloy) [inline types/react-relay/test/relay-experimental-tests.tsx:1102]:
```suggestion
        const entrypointB: EntryPoint<typeof RootEntryPointComponent, { author: string }> = {
```

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (alloy) [inline types/react-relay/test/relay-experimental-tests.tsx:1051]: Why is this `as any` cast necessary?

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (alloy) [inline types/react-relay/test/relay-experimental-tests.tsx:1092]: Ditto.

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (alloy) [inline types/react-relay/test/relay-experimental-tests.tsx:1109]: Ditto.

**Review by alloy — COMMENTED** [review]
As discussed, this is too large to go through with a comb, especially in the absence of tests for everything, but I like what I see when skimming and the new tests related to entry points are great (thanks!)

So let’s close the loop on some of the comments and then merge and iterate 👍

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (maraisr) [reply types/react-relay/test/relay-experimental-tests.tsx:1109]: Currently in OSS relay-compiler we don't have the types or generated files for EntryPoint params. But its a `$parameters` file. So im not sure what to type it as. But the current `ConcreteRequest` works.

Maybe since i do accept that as a type, this `as any` isnt required. Ill remove, and make sure tests pass.

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (maraisr) [reply types/react-relay/test/relay-experimental-tests.tsx:1092]: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/48638#discussion_r510080943

File: types/react-relay/test/relay-experimental-tests.tsx
Comment (maraisr) [reply types/react-relay/test/relay-experimental-tests.tsx:1051]: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/48638#discussion_r510080943

File: types/react-relay/lib/relay-experimental/helpers.d.ts
Comment (maraisr) [reply types/react-relay/lib/relay-experimental/helpers.d.ts:8]: Its got to do with that whole `$fragmentRefs`, `$refs` thing. So think once this gets merged, and I go through relay-runtime (much like this pr). Ill get a global picture of how things are meant to hang together and correct it then. For now this was needed to make this clean when authoring.

File: types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts
Comment (maraisr) [reply types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts:64]: Yeah, and the interface means we can do other fancy typescript things. `type` keyword has some weirdness around type matching, and discriminated union things.

File: types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts
Comment (maraisr) [reply types/react-relay/lib/relay-experimental/EntryPointTypes.d.ts:219]: I think its okay, it was just making types flow through when this is set. But its a really hard thing to type anyway. Think we can leave it for now, and revisit later.

File: types/entria__relay-experimental/index.d.ts
Comment (sibelius) [inline types/entria__relay-experimental/index.d.ts:5]: we can just remove entria__relay-experimental, we are not using anymore

**Review by sibelius — APPROVED** [review]

File: types/entria__relay-experimental/index.d.ts
Comment (maraisr) [reply types/entria__relay-experimental/index.d.ts:5]: Did you mind doing that? I don't want to delay this pr any longer.

Comment (alloy) [reply types/entria__relay-experimental/index.d.ts:5]: So was this bump to deal with requirements of `entria__relay-experimental`, because in that case I’d prefer not to bump.

**Review by edvinerikson — APPROVED** [review]

**Comment by elibarzilay** [comment]
@alloy: not clear to me whether your objection should hold this off or not.

File: types/entria__relay-experimental/index.d.ts
Comment (maraisr) [reply types/entria__relay-experimental/index.d.ts:5]: The bump needs to happen because ts version before 3.7 doesn't support recursive types, and we need that for the EntryPoints

Comment (alloy) [reply types/entria__relay-experimental/index.d.ts:5]: Cool, got it 👍

**Comment by alloy** [comment]
@elibarzilay Now that my question has been answered it's good to go.

**Pull request merged**
